### DISCUSSION TITLE
Having difficulties dynamically updating ProgressBar

### DISCUSSION THREAD

**Discussion by rlad78** [Textualize/textual#2550] [Q&A] [answered]
Hi, I'm very new to Textual and moderately experienced with Python async. I may have a misconception or a missing piece of information that's causing me a great deal of headache...

I'm trying to write a progress bar that updates dynamically based on work done in a separate queue. I've got a callback system set up where work being completed should trigger either an increase in the bar's total or the bar's progress. It functions something like this:

```python
class Progress(Screen):
    def compose(self) -> ComposeResult:
        with Center():
            with Middle():
                yield ProgressBar(name="SNMP Tasks", id="snmp_bar")
                yield ProgressBar(name="Connect Tasks", id="connect_bar")
                yield ProgressBar(name="SSH Tasks", id="ssh_bar")

    def on_mount(self) -> None:
        add_queue_callback_func(self._bar_progress, queue_completions)
        add_queue_callback_func(self._bar_total, queue_additions)
        self.run_worker(do_async_queue_things())

    def _bar_progress(self, key: SubKey) -> None:
        if key[1] == "snmp":
            logger.critical("GOT AN SNMP HIT")
            self.query_one("#snmp_bar", ProgressBar).advance()
        ...

    def _bar_total(self, key: SubKey, token: RequestToken) -> None:
        def inc_total(bar: ProgressBar) -> None:
            bar_total = bar.total
            if bar_total is not None:
                new_total = int(bar_total) + 1
            else:
                new_total = 1
            bar.update(total=new_total)

        if key[1] == "snmp":
            logger.critical("GOT AN SNMP TOTAL")
            inc_total(self.query_one("#snmp_bar", ProgressBar))
        ...
```

The three ProgressBars will render when I create and push the Screen. The `_bar_progress` and `_bar_total` callback functions get called plenty of times, according to my logs. However, I see nothing happen as a result of the ProgressBar's `advance()` or `update()` being called. Not even seeing a "Changed" event appear in the Textual debugging console.

Does anyone have an idea on what I might be doing wrong here?

**Comment by davep** [comment]
It's a bit tricky to say exactly what might be the issue here given that this is only a snippet of some code. Do you think you might be able to reproduce the issue as some standalone code anyone can run?

Meanwhile though, given that it seems you're writing something that involves doing work parallel to updating the display, perhaps Textual's [worker API](https://textual.textualize.io/guide/workers/) would be helpful here?

**Reply by rlad78** [reply]
After writing up a "demo" script with similar functionality to my app, I found the issue...

I set my `_bar_progress` increment method as the callback for both queue item completions _and_ additions. This meant that I was increasing the ProgressBar's `progress` value, but never setting or increasing the `total` value.

Out of curiosity, would the `validate_progress()` method have helped me catch the issue earlier? From the documentation, I'm a bit confused by the word "clamp." Does it floor the progress value under the ProgressBar's total? Or will it raise an exception if progress > total?

**Reply by davep** [reply]
Looking at the code, `validate_progress` calls on [`clamp`](https://textual.textualize.io/api/geometry/#textual.geometry.clamp) to force the value between the lower and upper bounds; no exception should be raised.

**Comment by rlad78** [comment, marked as answer]
Issue was due to my own programming error. Closing discussion. Thanks davep!

**Reply by davep** [reply]
Glad to hear you got to the bottom of it in the end. :-)
